src/main/java/htsjdk/variant/variantcontext/VariantContext.java

```diff
@@ -1638,106 +1651,221 @@ public boolean isFullyDecoded() {
+                isFormat ? "FORMAT" : "INFO", id
+            ));
+        }
+        if (line == null) {
+            // Try to find a matching standard header line
+            line = isFormat
+                ? VCFStandardHeaderLines.getFormatLine(id, false)
+                : VCFStandardHeaderLines.getInfoLine(id, false);
+        }
+        if (line == null) {
+            final String description = "Auto-generated unbounded string "
+                + (isFormat ? "FORMAT" : "INFO")
+                + " header line for " + id;
+            // Decode values as strings, and allow an unbounded number of values
+            line = isFormat
+                ? new VCFFormatHeaderLine(id, VCFHeaderLineCount.UNBOUNDED, VCFHeaderLineType.String, description)
+                : new VCFInfoHeaderLine(id, VCFHeaderLineCount.UNBOUNDED, VCFHeaderLineType.String, description);
+        }
+        return line;
+    }
+
+    // ---------------------------------------------------------------------------------------------------------
+    //
+    // Version awareness related methods
+    //
+    // ---------------------------------------------------------------------------------------------------------
+
+    public VCFHeaderVersion getVersion() {
+        return version;
+    }
+
+    /**
+     * Ensure that this VariantContext is compatible with the VCF version of the given header. Discrepancies can occur
+     * in the encoding of the GP key or percent encoding of strings between versions. Returns a new VariantContext
+     * if the version of this VariantContext differs from the version of the header
+     * @param header the header to ensure compatibility with
+     * @return the current header if the version of this VariantContext matches the version of the header, otherwise
+     * a new, fully decoded VariantContext, whose attributes are compatible with the header's version
+     */
+    public VariantContext makeCompatibleWithHeaderVersion(final VCFHeader header) {
+        final VCFHeaderVersion headerVersion = header.getVCFHeaderVersion();
+        if (this.version.equals(headerVersion)) {
+            return this;
+        } else {
+            // This will ensure that the VariantContext is decoded in a percent-encoding aware way
+            final VariantContext newVC = this.fullyDecode(header, true);
+            newVC.version = headerVersion;
+
+            final Object oldGP = newVC.getAttribute(VCFConstants.GENOTYPE_POSTERIORS_KEY);
+            // We need to special-case GP because there is a discrepancy in the scale used to record
+            // its values between pre-4.3 and 4.3+ VCF. Pre-4.3 GP is phred scale encoded while
+            // 4.3+ GP is a linear probability, bringing it in line with other standard keys that
+            // use the P suffix (c.f. VCF 4.3 spec section 7.2).
+            if (headerVersion.isAtLeastAsRecentAs(VCFHeaderVersion.VCF4_3) && oldGP != null) {
```

Comment:
Shouldn't this also check the VC version ? We shouldn't need to do this if its already 4.3 ?

Reply:
The check would've been redundant before because we were checking whether the version was the same, but with the other change I added it in